libpgcommon/lwgeom_transform.c

```diff
@@ -666,147 +825,153 @@ void SetPROJ4LibPath(void)
-*/
-static void
-SetSpatialRefSysSchema(FunctionCallInfo fcinfo)
-{
-	char *nsp_name;
-
-	/* Schema info is already cached, we're done here */
-	if (spatialRefSysSchema) return;
-
-	/* For some reason we have a hobbled fcinfo/flinfo */
-	if (!fcinfo || !fcinfo->flinfo) return;
-
-	nsp_name = get_namespace_name(get_func_namespace(fcinfo->flinfo->fn_oid));
-	/* early exit if we cannot lookup nsp_name, cf #4067 */
-	if (!nsp_name) return;
-
-	elog(DEBUG4, "%s located %s in namespace %s", __func__, get_func_name(fcinfo->flinfo->fn_oid), nsp_name);
-	spatialRefSysSchema = MemoryContextStrdup(CacheMemoryContext, nsp_name);;
-	return;
-}
 
 int
-GetProjectionsUsingFCInfo(FunctionCallInfo fcinfo, int srid1, int srid2, projPJ *pj1, projPJ *pj2)
+GetPJUsingFCInfo(FunctionCallInfo fcinfo, int srid_from, int srid_to, PJ** pj)
 {
-	Proj4Cache *proj_cache = NULL;
+	PROJPortalCache *proj_cache = NULL;
 
 	/* Set the search path if we haven't already */
-	SetPROJ4LibPath();
+	SetPROJLibPath();
 
 	/* Look up the spatial_ref_sys schema if we haven't already */
 	SetSpatialRefSysSchema(fcinfo);
 
 	/* get or initialize the cache for this round */
-	proj_cache = GetPROJ4Cache(fcinfo);
-	if ( !proj_cache )
+	proj_cache = GetPROJSRSCache(fcinfo);
+	if (!proj_cache)
 		return LW_FAILURE;
 
 	/* Add the output srid to the cache if it's not already there */
-	if (!IsInPROJ4Cache(proj_cache, srid1))
-		AddToPROJ4Cache(proj_cache, srid1, srid2);
-
-	/* Add the input srid to the cache if it's not already there */
-	if (!IsInPROJ4Cache(proj_cache, srid2))
-		AddToPROJ4Cache(proj_cache, srid2, srid1);
+	if (!IsInPROJSRSCache(proj_cache, srid_from, srid_to))
+		AddToPROJSRSCache(proj_cache, srid_from, srid_to);
 
 	/* Get the projections */
-	*pj1 = GetProjectionFromPROJ4Cache(proj_cache, srid1);
-	*pj2 = GetProjectionFromPROJ4Cache(proj_cache, srid2);
+	*pj = GetProjectionFromPROJCache(proj_cache, srid_from, srid_to);
 
 	return LW_SUCCESS;
 }
 
-int
-spheroid_init_from_srid(FunctionCallInfo fcinfo, int srid, SPHEROID *s)
-{
-	projPJ pj1, pj2;
-#if POSTGIS_PROJ_VERSION >= 48
-	double major_axis, minor_axis, eccentricity_squared;
-#endif
 
-	if ( GetProjectionsUsingFCInfo(fcinfo, srid, srid, &pj1, &pj2) == LW_FAILURE)
-		return LW_FAILURE;
-
-	if ( ! pj_is_latlong(pj1) )
-		return LW_FAILURE;
-
-#if POSTGIS_PROJ_VERSION >= 48
-	/* For newer versions of Proj we can pull the spheroid paramaeters and initialize */
-	/* using them */
-	pj_get_spheroid_defn(pj1, &major_axis, &eccentricity_squared);
-	minor_axis = major_axis * sqrt(1-eccentricity_squared);
-	spheroid_init(s, major_axis, minor_axis);
+static int
+proj_pj_is_latlong(const PJ* pj)
+{
+#if POSTGIS_PROJ_VERSION < 60
+	return pj_is_latlong(pj->pj_from);
 #else
-	/* For old versions of Proj we cannot lookup the spheroid parameters from the API */
-	/* So we use the WGS84 parameters (boo!) */
-	spheroid_init(s, WGS84_MAJOR_AXIS, WGS84_MINOR_AXIS);
+	PJ_TYPE pj_type = proj_get_type(proj_get_source_crs(NULL, pj));
```

Comment:
This will crash if `proj_get_source_crs` returns NULL:

When running regress/core/geography:
```
#0  0x00007ff96c529d7f in raise () from /usr/lib/libc.so.6
#1  0x00007ff96c514672 in abort () from /usr/lib/libc.so.6
#2  0x00007ff96c514548 in __assert_fail_base.cold.0 () from /usr/lib/libc.so.6
#3  0x00007ff96c522396 in __assert_fail () from /usr/lib/libc.so.6
#4  0x00007ff75bb8fb05 in proj_get_type (obj=<optimized out>) at iso19111/c_api.cpp:904
#5  0x00007ff75bd5764c in proj_pj_is_latlong (pj=0x557002aacea0) at lwgeom_transform.c:874
#6  spheroid_init_from_srid (fcinfo=<optimized out>, srid=<optimized out>, s=0x7fffa99d5828) at lwgeom_transform.c:939
#7  0x00007ff75bd1f8c8 in geography_distance (fcinfo=0x5570027d3ec0) at geography_measurement.c:223
#8  0x00005570003243e0 in ExecInterpExpr (state=<optimized out>, econtext=<optimized out>, isnull=0x7fffa99d599f) at execExprInterp.c:678
#9  0x0000557000356bf9 in ExecEvalExprSwitchContext (state=0x5570027d3de8, econtext=0x5570027d3b10, isNull=0x3242a5f189c500da)
    at ../../../src/include/executor/executor.h:303
#10 ExecProject (projInfo=0x5570027d3de0) at ../../../src/include/executor/executor.h:337
#11 ExecResult (pstate=<optimized out>) at nodeResult.c:136
#12 0x000055700032c200 in ExecProcNode (node=<optimized out>) at ../../../src/include/executor/executor.h:237
#13 ExecutePlan (estate=0x5570027d37f0, planstate=0x5570027d3a00, use_parallel_mode=<optimized out>, operation=CMD_SELECT, 
    sendTuples=<optimized out>, numberTuples=1, direction=ForwardScanDirection, dest=0x5570027d2510, execute_once=<optimized out>)
    at execMain.c:1723
```

Reply:
It also leaks memory:
```
 * @return Object that must be unreferenced with proj_destroy(), or NULL
 * in case of error, or missing source CRS.
 */
PJ *proj_get_source_crs(PJ_CONTEXT *ctx, const PJ *obj) {
```